### PR TITLE
Tw 2046 group name validation

### PR DESCRIPTION
**Pull request by KhaledNjim**
## Ticket
- #2046 

## Solution
 Added validation for group name

## Impact description
**If this is not a bug, please explain how the changes affect the project**

## Resolved
- Web:

https://github.com/user-attachments/assets/2cf86cf2-187a-47ab-b7f4-a24f003a4a71

- Android:
- IOS:

### CODE REVIEW

File: lib/domain/model/verification/composite_name_validator.dart [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import 'package:dartz/dartz.dart';
+import 'package:fluffychat/app_state/failure.dart';
+import 'package:fluffychat/app_state/success.dart';
+import 'package:fluffychat/domain/app_state/validator/verify_name_view_state.dart';
+import 'package:fluffychat/domain/model/verification/new_name_request.dart';
+import 'package:fluffychat/domain/model/verification/validator.dart';
+import 'package:fluffychat/domain/model/extensions/list_validator_extension.dart';
+
+class CompositeNameValidator extends Validator<NewNameRequest> {
```
Comment (hoangdat) [inline lib/domain/model/verification/composite_name_validator.dart:9]: should write test for this `validator`

File: lib/domain/usecase/verify_name_interactor.dart [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import 'package:dartz/dartz.dart';
+import 'package:fluffychat/app_state/failure.dart';
+import 'package:fluffychat/app_state/success.dart';
+import 'package:fluffychat/domain/app_state/validator/verify_name_view_state.dart';
+import 'package:fluffychat/domain/model/verification/composite_name_validator.dart';
+import 'package:fluffychat/domain/model/verification/new_name_request.dart';
+import 'package:fluffychat/domain/model/verification/validator.dart';
+
+class VerifyNameInteractor {
```
Comment (hoangdat) [inline lib/domain/usecase/verify_name_interactor.dart:9]: unit test this

**Review by hoangdat — APPROVED** [review]

**Comment by hoangdat** [comment]
@nqhhdev please review this PR

**Comment by nqhhdev** [comment]
Pls update for edit group also

Screenshot 2024-11-18 at 14 58 05

File: lib/domain/app_state/validator/verify_name_view_state.dart [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import 'package:fluffychat/presentation/state/failure.dart';
+import 'package:fluffychat/presentation/state/success.dart';
+
+class VerifyNameViewState extends UIState {}
```
Comment (nqhhdev) [inline lib/domain/app_state/validator/verify_name_view_state.dart:4]: `VerifyNameViewState` => `Verified NameViewState` or `VerifyNameSuccessViewState`

File: lib/domain/model/verification/name_with_space_only_validator.dart [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import 'package:fluffychat/domain/app_state/validator/verify_name_view_state.dart';
+import 'package:fluffychat/domain/exception/verify_name_exception.dart';
+import 'package:fluffychat/domain/model/verification/new_name_request.dart';
+import 'package:fluffychat/domain/model/verification/validator.dart';
+
+class NameWithSpaceOnlyValidator extends Validator<NewNameRequest> {
+  @override
+  Either<Failure, Success> validate(NewNameRequest value) {
+    if (value.value != null &&
+        value.value!.isNotEmpty &&
```
Comment (nqhhdev) [inline lib/domain/model/verification/name_with_space_only_validator.dart:13]: `isNotEmpty` or `isEmpty` ?

File: assets/l10n/intl_en.arb [hunk 5]
```diff
@@ -3097,5 +3097,6 @@
   "disable_notifications": "Disable notifications",
   "logoutDialogWarning": "You will lose access to encrypted messages. We recommend that you enable chat backups before logging out",
   "copyNumber": "Copy number",
-  "callViaCarrier": "Call via Carrier"
+  "callViaCarrier": "Call via Carrier",
+  "this_field_cannot_be_blank":"This field cannot be blank"
```
Comment (nqhhdev) [inline assets/l10n/intl_en.arb:3101]:
```suggestion
  "thisFieldCannotBeBlank":"This field cannot be blank"
```

File: assets/l10n/intl_en.arb [hunk 6]
```diff
@@ -3098,5 +3098,6 @@
   "logoutDialogWarning": "You will lose access to encrypted messages. We recommend that you enable chat backups before logging out",
   "copyNumber": "Copy number",
   "callViaCarrier": "Call via Carrier",
-  "scanQrCodeToJoin": "Installation of the mobile application will allow you to contact people from your phone's address book, your chats will be synchronised between devices"
+  "scanQrCodeToJoin": "Installation of the mobile application will allow you to contact people from your phone's address book, your chats will be synchronised between devices",
+  "thisFieldCannotBeBlank":"This field cannot be blank"
```
Comment (nqhhdev) [inline assets/l10n/intl_en.arb:3102]:
```suggestion
  "thisFieldCannotBeBlank": "This field cannot be blank"
```

File: lib/domain/model/verification/name_with_space_only_validator.dart (hunk 4 above)
Comment (nqhhdev) [reply lib/domain/model/verification/name_with_space_only_validator.dart:13]: Could you help me explain for this case?

Comment (KhaledNjim) [reply lib/domain/model/verification/name_with_space_only_validator.dart:13]: To verify if the given string contains only spaces,  it needs to be not null and not empty (before trim) but after trimming the spaces it needs to be empty

**Comment by nqhhdev** [comment]
IMO: Side effect, pls check it

https://github.com/user-attachments/assets/094b50f3-0119-4d85-9f83-8f2c5a9da1e0

**Comment by KhaledNjim** [comment]
> IMO: Side effect, pls check it
> 
>  Screen.Recording.2024-11-19.at.15.37.48.mov

i can reproduce this on older version, this happens when the name group is empty so the text displayed is just  a placeholder returned by `getLocalizedDisplayname()`

https://github.com/user-attachments/assets/7361e8a7-69f2-4b20-b2bd-dd13c4cf83d3


image

**Comment by nqhhdev** [comment]
> > IMO: Side effect, pls check it
>
> i can reproduce this on older version, this happens when the name group is empty so the text displayed is just a placeholder returned by `getLocalizedDisplayname()` so this is unrelated to this pr
>

If empty chat I think no need to warning. Only warning after user edited

File: lib/pages/chat_details/chat_details_edit.dart [hunk 7]
```diff
@@ -443,6 +450,24 @@ class ChatDetailsEditController extends State<ChatDetailsEdit>
...
+            (failure) {
+              if (failure is VerifyNameFailure) {
+                return failure.getMessage(context);
+              } else {
+                return null;
+              }
+            },
+            (success) => null,
+          );
+  }
```
Comment (nqhhdev) [inline lib/pages/chat_details/chat_details_edit.dart:469]:
```suggestion
  String? getErrorMessage(String content) {
    if (content == room?.name) {
      return null;
    }

    final result = verifyNameInteractor.execute(
      content,
      [EmptyNameValidator(), NameWithSpaceOnlyValidator()],
    );

    return result.fold(
      (failure) =>
          failure is VerifyNameFailure ? failure.getMessage(context) : null,
      (success) => null,
    );
  }
```

Comment (nqhhdev) [reply lib/pages/chat_details/chat_details_edit.dart]: Hi @KhaledNjim why resolved this conversation?

Comment (KhaledNjim) [reply lib/pages/chat_details/chat_details_edit.dart]: its fixed

Comment (nqhhdev) [reply lib/pages/chat_details/chat_details_edit.dart]: What's different? What I mean is don't use ternary operators like that for many conditions. PLSSSSSS

**Review by hoangdat — APPROVED** [review]

**Review by hoangdat — APPROVED** [review]

**Pull request merged**